src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -2484,166 +2487,71 @@ public int EndReceiveFrom(IAsyncResult asyncResult, ref EndPoint endPoint)
+        public Socket EndAccept(IAsyncResult asyncResult)
         {
             ThrowIfDisposed();
+            return TaskToApm.End<Socket>(asyncResult);
+        }
 
+        // This method provides support for legacy BeginAccept methods that take a "receiveSize" argument and
+        // allow data to be received as part of the accept operation.
+        // There's no direct equivalent of this in the Task APIs, so we mimic it here.
+        private Task<(Socket s, byte[] buffer, int bytesReceived)> AcceptAndReceiveHelperAsync(Socket? acceptSocket, int receiveSize)
+        {
             // Validate input parameters.
             if (receiveSize < 0)
             {
                 throw new ArgumentOutOfRangeException(nameof(receiveSize));
             }
 
-            // Set up the async result with flowing.
-            AcceptOverlappedAsyncResult asyncResult = new AcceptOverlappedAsyncResult(this, state, callback);
-            asyncResult.StartPostingAsyncOp(false);
+            Task<Socket> acceptTask = AcceptAsync(acceptSocket);
+            acceptTask.ThrowIfFailedSynchronously();
 
-            // Start the accept.
-            if (_rightEndPoint == null)
+            static async Task<(Socket s, byte[] buffer, int bytesReceived)> Run(Task<Socket> acceptTask, int receiveSize)
             {
-                throw new InvalidOperationException(SR.net_sockets_mustbind);
-            }
-
-            if (!_isListening)
-            {
-                throw new InvalidOperationException(SR.net_sockets_mustlisten);
-            }
-
-            SafeSocketHandle? acceptHandle;
-            asyncResult.AcceptSocket = GetOrCreateAcceptSocket(acceptSocket, false, nameof(acceptSocket), out acceptHandle);
-
-            if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, $"AcceptSocket:{acceptSocket}");
-            if (SocketsTelemetry.Log.IsEnabled()) SocketsTelemetry.Log.AcceptStart(_rightEndPoint);
-
-            int socketAddressSize = GetAddressSize(_rightEndPoint);
-            SocketError errorCode;
-            try
-            {
-                errorCode = SocketPal.AcceptAsync(this, _handle, acceptHandle, receiveSize, socketAddressSize, asyncResult);
-            }
-            catch (Exception ex)
-            {
-                if (SocketsTelemetry.Log.IsEnabled())
+                Socket s = await acceptTask.ConfigureAwait(false);
+                byte[] buffer;
+                int bytesReceived;
+                if (receiveSize == 0)
                 {
-                    SocketsTelemetry.Log.AfterAccept(SocketError.Interrupted, ex.Message);
+                    buffer = Array.Empty<byte>();
+                    bytesReceived = 0;
+                }
+                else
+                {
+                    buffer = new byte[receiveSize];
+                    bytesReceived = await s.ReceiveAsync(buffer, SocketFlags.None).ConfigureAwait(false);
```

Comment:
I believe it's possible the AcceptEx could previously have completed both the accept/receive synchronously, with errors from the receive aspect propagated synchronously.  Now that receive is a separate operation inside of an async method, the workaround this PR is trying to utilize doesn't achieve its goal of making everything that could have failed synchronously before still fail synchronously.